unittests/finality_tests.cpp

```diff
@@ -0,0 +1,739 @@
+         produce_and_push_block();
+         process_bob_vote();
+         if (!node1_lib_advancing() || !node2_lib_advancing() || !node3_lib_advancing()) {
+            return false;
+         }
+      }
+
+      return true;
+   }
+
+   std::vector<eosio::chain::vote_message> bob_votes;
+
+private:
+   bool lib_advancing(uint32_t curr_lib_num, uint32_t& prev_lib_num) {
+      auto advancing = curr_lib_num > prev_lib_num;
+      // update pre_lib_num for next time check
+      prev_lib_num = curr_lib_num;
+      return advancing;
+   }
+
+   eosio::testing::tester   node1;
+   eosio::testing::tester   node2;
+   eosio::testing::tester   node3;
+   uint32_t                 node1_prev_lib_bum {0};
+   uint32_t                 node2_prev_lib_bum {0};
+   uint32_t                 node3_prev_lib_bum {0};
+   std::vector<eosio::chain::vote_message> carol_votes;
+
+   void setup_node(eosio::testing::tester& node, eosio::chain::account_name local_finalizer) {
+      std::vector<eosio::chain::account_name> account_names = { "alice"_n, "bob"_n, "carol"_n };
+
+      node.produce_block();
+      node.create_accounts(account_names);
+      node.produce_block();
+
+      // activate hotstuff
+      eosio::testing::base_tester::finalizer_policy_input policy_input = {
+         .finalizers       = { {.name = "alice"_n, .weight = 1},
+                               {.name = "bob"_n, .weight = 1},
+                               {.name = "carol"_n, .weight = 1}},
+         .threshold        = 2,
+         .local_finalizers = {local_finalizer}
+      };
+      node.set_finalizers(policy_input);
+      auto block = node.produce_block();
+
+      // this block contains the header extension for the instant finality
+      std::optional<eosio::chain::block_header_extension> ext = block->extract_header_extension(eosio::chain::instant_finality_extension::extension_id());
+      BOOST_TEST(!!ext);
+      std::optional<eosio::chain::finalizer_policy> fin_policy = std::get<eosio::chain::instant_finality_extension>(*ext).new_finalizer_policy;
+      BOOST_TEST(!!fin_policy);
+      BOOST_TEST(fin_policy->finalizers.size() == account_names.size());
+      BOOST_TEST(fin_policy->generation == 1);
+   }
+};
+
+// verify LIB advances with 2 finalizers voting and for a long time
+BOOST_AUTO_TEST_CASE(two_votes_large_number_blocks) { try {
+   tester_cluster cluster;
+
+   for (auto i = 0; i < 1000; ++i) {
+      // node1 produces a block and pushes to node2 and node3
+      cluster.produce_and_push_block();
+      // process the Bob's votes only
+      cluster.process_bob_vote();
+
+      // all nodes advance LIB
+      BOOST_REQUIRE(cluster.node1_lib_advancing());
+      BOOST_REQUIRE(cluster.node2_lib_advancing());
+      BOOST_REQUIRE(cluster.node3_lib_advancing());
+   }
+} FC_LOG_AND_RETHROW() }
+
+// verify LIB advances with all of the three finalizers voting
+BOOST_AUTO_TEST_CASE(all_votes) { try {
+   tester_cluster cluster;
+
+   for (auto i = 0; i < 50; ++i) {
```

Comment:
Don't see why this needs to be tested 50 times. 2 or 3 seems plenty. Same for other tests.

Reply:
Will reduce to 3 for this and others.